File: 2011_2012_sujet_2_simulation_de_lois/TPsimu2.py
"""TP Simulation 2 — 2011-2012 (suite)."""
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm

from plotrep import plotrep
from runif import runif
from rexp import rexp
from mystere import mystere

# %% 14. Exponential via CDF inversion
E = rexp(1000, 2)
print(np.mean(E))
plt.figure()
plt.hist(E, bins=20, density=True)
x = np.linspace(0, 4, 200)
plt.plot(x, 2 * np.exp(-2 * x))
plt.show()
plt.figure()
plotrep(E)
plt.plot(x, 1 - np.exp(-2 * x), 'r')
plt.show()

# %% 16. Geometric — mystere(p) generates Geometric(p)
geom = np.array([mystere(1/6) for _ in range(1000)])
print(np.mean(geom))
w = np.ones(1000) / 1000
xg = np.arange(1, 41)
plt.figure()
plt.hist(geom, bins=np.arange(0.5, max(geom)+1.5), weights=w)
plt.stem(xg, 1/6 * (5/6)**(xg - 1), 'r', markerfmt='ro', basefmt=' ')
plt.show()

# %% 17. Normal(10, 5)
Xn = 10 + np.random.randn(1000) * 5
plt.figure()
plt.hist(Xn, bins=20, density=True)
x = np.linspace(-10, 30, 200)
plt.plot(x, norm.pdf(x, 10, 5), 'r')
plt.show()
# Alternative
plt.figure()
plt.hist(Xn, bins=20, density=True)
plt.plot(x, norm.pdf(x, 10, 5), 'r')
plt.show()

# %% 18. Monte Carlo disk
X = 2 * np.random.rand(10000) - 1
Y = 2 * np.random.rand(10000) - 1
I = np.where(X**2 + Y**2 < 1)[0]
Xd, Yd = X[I], Y[I]
plt.figure()
plt.plot(Xd, Yd, 'd', ms=1)
t = np.linspace(0, 2*np.pi, 200)
plt.plot(np.cos(t), np.sin(t), 'r')
plt.axis('equal')
plt.show()

plt.figure(); plt.hist(Xd, bins=15, density=True); plt.title('Xd not uniform'); plt.show()

# %% Non-uniform polar sampling
r = np.random.rand(10000)
t2 = 2 * np.pi * np.random.rand(10000)
plt.figure()
plt.plot(r * np.cos(t2), r * np.sin(t2), '.', ms=1)
plt.show()

# %% Cardioid rejection
Abs1 = runif(10000, -1/4, 2)
Ord1 = runif(10000, -3*np.sqrt(3)/4, 3*np.sqrt(3)/4)
R = np.sqrt(Abs1**2 + Ord1**2)
T = np.arctan2(Ord1, Abs1)
I = np.where(R - np.cos(T) < 1)[0]
Cardx, Cardy = Abs1[I], Ord1[I]
plt.figure()
plt.plot(Cardx, Cardy, 'd', ms=1)
t3 = np.linspace(0, 2*np.pi, 200)
plt.plot((1+np.cos(t3))*np.cos(t3), (1+np.cos(t3))*np.sin(t3), 'r')
plt.show()

File: 2011_2012_sujet_2_simulation_de_lois/TPsimu2.m
%% 14 Par inversion de fonction de repartition : loi exponentielle
E=rexp(1000,2);
mean(E)
histo(E,20,0,1)
hold on
fplot(@(x)2*exp(-2.*x),[0 4])
hold off
figure
plotrep(E);
hold on
fplot(@(x)1-exp(-2*x),[0 4],'r')
hold off
%% 16 Mystere permet de generer une loi geometrique de parametre p
geom=zeros(1000,1);
for i=1:1000
   geom(i)=mystere(1/6);
end
mean(geom)
histo(geom,max(geom)-min(geom)+1,1,1)
hold on;
xg=[1:40]; stem(xg, 1/6*(5/6).^(xg-1),'r')
hold off;
%% 17
Xn=10+randn(1000,1)*5;
histo(Xn,20,0,1)
hold on
fplot(@(x)1./sqrt(2*pi*25)*exp(-(x-10).^2/2/25),[-10 30],'r')
hold off
% ou bien
histo(Xn,20,0,1)
hold on
fplot(@(x)dnorm(x,10,5),[-10 30],'r')
hold off
type rnorm
%%
%% 18
clf
X=2*rand(10000,1)-1;
Y=2*rand(10000,1)-1;
I=find((X.^2+Y.^2)<1);
Xd=X(I); Yd=Y(I);
plot(Xd,Yd,'d')
hold on
clear i %i a ete utilise precedemment 
t=[0:pi/30:2*pi]; plot(exp(i*t),'r')
% ou bien
t=[0:pi/30:2*pi]; plot(cos(t),sin(t),'r')
%ou bien
t=[0:pi/30:2*pi]; polar(t,ones(1,61),'r')
hold off
%%
figure;
histo(Xd,15,0,1);%Xd ne suit pas une loi uniforme.
%%
figure;
r=rand(10000,1);
t=2*pi*rand(10000,1);
plot(r.*exp(i*t),'.') % La loi n'est pas uniforme.
%%
Abs1=runif(10000,-1/4,2);
Ord1=runif(10000,-3*sqrt(3)/4, 3*sqrt(3)/4);
[T,R]=cart2pol(Abs1,Ord1);
I=find(R-cos(T)<1);
Cardx=Abs1(I); Cardy=Ord1(I);
plot(Cardx,Cardy,'d')
hold on
t=[0:pi/30:2*pi]; plot((1+cos(t)).*exp(i*t),'r')
%ou bien
t=[0:pi/30:2*pi];polar(t,1+cos(t),'r')
hold off